rgw: improve convenience for key operate.

guce opened the pull request:
**1.key-type assignments based on context**
* In user    operate context, key-type assignment to `KEY_TYPE_S3`
* In subuser operate context, key-type assignment to `KEY_TYPE_SWIFT`
* In key     operate context, if key-type wasn't specified by previous process, key-type assignments based on `user type`.

**2.fix `RGWSubUserPool::add()`**
* When create subuser generate secret by default

**3.fix `RGWAccessKeyPool::generate_key()`**
* Avoid empty secret when use --access-key para
* Avoid wrong key's username when create user and subuser at the same time

Signed-off-by: Ce Gu <guce@h3c.com>

guce commented:
# I have made the following test:

**1.user create**
`radosgw-admin user create --uid=111 --display-name=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | ok | ok
. | specified | specified | ok | ok
. | specified | generate | ok | ok
. | generate | specified | ok | ok
. | generate | generate | ok | ok



**2.subuser create**
`radosgw-admin subuser create --uid=111 --subuser=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | Failed, s3 key and NULL secret | ok
swift | . | . | Failed, NULL secret | ok
swift | . | specified | ok | ok
swift | . | generate | ok | ok



**3.user subuser same time create**
`radosgw-admin user create --uid=111 --display-name=111 --subuser=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | Failed, two s3 key and key's user was subuser | ok



**4.key create for user**
`radosgw-admin key create --uid=111 --gen-access-key`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | generate | . | ok | ok
. | generate | generate | ok | ok
. | generate | specified | ok | ok
. | specified | . | Failed, NULL secret | ok
. | specified | generate | ok | ok
. | specified | specified | ok | ok



**5.key create subuser**
`radosgw-admin key create --uid=111 --subuser=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | Failed, error message | ok
swift | . | . | Failed, NULL secret | ok
swift | . | generate | ok | ok
swift | . | specified | ok | ok



**6.key rm for user**
`radosgw-admin key rm --uid=111 --access-key=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | specified | . | ok | ok



**7.key rm for subuser**
`radosgw-admin key rm --uid=111 --subuser=111`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | Failed, error message | ok
swift | . | . | ok | ok


**8.subuser rm**
`radosgw-admin subuser rm --uid=111 --subuser=111 --purge-keys`

Keytype | Access | Secret | Original | Modified
------- | ------ | ------ | -------- | --------
. | . | . | Failed, key still exist | ok
swift | . | . | ok | ok

guce commented:
@yehudasa Hi, could you please take a look, thanks.

yehudasa commented:
@guce see my comments. Also, have you created an automated script that you tested it? It'd be great incorporating it into our test suite

guce commented:
@yehudasa Your comment is very important, and I also want to know the expected specifications about key operate. Currently I test this manually, if you can give me same example, I would like to write same script.

guce commented:
@yehudasa I expect continue discuss with you, which will allow me to learn how to modify the code.

yehudasa commented:
@guce sorry for not being responsive, got sucked into other stuff, will try to get to it early next week. Thanks!

guce commented:
@yehudasa ok, happy weekend.

guce commented:
@yehudasa

guce commented:
@yehudasa hi~~~~~

guce commented:
@dachary @smithfarm
Can someone look at this issue it, thank you

dachary commented:
@guce I would like to help but I don't know enough about rgw. It is not necessary to ping if there is an "Assignee" field set, your pull request is in the TODO list already. Unless a month pass and nothing happens, of course ;-)

yehudasa commented:
@guce hopefully @oritwas will be able to take a look at that soon. Sorry for not taking a look at it last week.

oritwas commented on src/rgw/rgw_user.cc:
```diff
@@ -1238,6 +1243,8 @@ int RGWSubUserPool::check_op(RGWUserAdminOpState& op_state,
     return -EINVAL;
   }
 
+  op_state.set_key_type(KEY_TYPE_SWIFT);
```
I suggest setting the key type only if it not already set

oritwas commented on src/rgw/rgw_user.cc:
```diff
@@ -1756,6 +1763,8 @@ int RGWUser::check_op(RGWUserAdminOpState& op_state, std::string *err_msg)
     return -EINVAL;
   }
 
+  op_state.set_key_type(KEY_TYPE_S3);
```
same here , setting the key type only if it not set

guce commented:
@oritwas ok, I'll modify it early next week.

guce commented:
@oritwas 

There is only a small problem left. If set keytype only when it not set in subuser context.

When create user and subuser at the same time, it will create a s3 key for subuser.
e.g(`radosgw-admin user create --uid=111 --display-name=111 --subuser=111`)
because the keytype will set to s3 in user create context

Should we need to consider this?

oritwas commented:
yes I guess for subuser we will need to set the key type to swift always

guce commented:
@oritwas

Hi, I have modified, please check.
I added a var `key_type_setbycontext` to indicate whether the key_type set by context. This can keep the 

caller's flexibility.

oritwas commented on src/rgw/rgw_user.cc:
```diff
@@ -878,12 +882,13 @@ int RGWAccessKeyPool::generate_key(RGWUserAdminOpState& op_state, std::string *e
     new_key.subuser = op_state.get_subuser();
 
-  if (!gen_secret) {
+  if (!op_state.get_secret_key().empty()) {
```
should be op_state.will_gen_secret() here

guce commented:
@oritwas 

If use `op_state.will_gen_secret()`, when specifed `--access-key` but no `--gen-secret` nor `--secret` like below command, it will create a key pair which secret was empty.
`radosgw-admin key create --uid=111 --access-key=11111`
`radosgw-admin subuser create --uid=111 --subuser=111 --key-type=swift`

Should we need to consider this?

oritwas commented:
the admin command should not allow this , can you fix it?

guce commented:
@oritwas 

I submit the code below, to avoid empty secret.
if 'get_secret_key()' not empty, use the specified value, otherwise generate a secret by default.

```
RGWAccessKeyPool::generate_key()

  if (!op_state.get_secret_key().empty()) {
    key = op_state.get_secret_key();
  } else {
    char secret_key_buf[SECRET_KEY_LEN + 1];

    ret = gen_rand_alphanumeric_plain(g_ceph_context, secret_key_buf, sizeof(secret_key_buf));
    if (ret < 0) {
      set_err_msg(err_msg, "unable to generate secret key");
      return ret;
    }

    key = secret_key_buf;
  }
```

oritwas commented:
In RGWSubUserPool::add we set the get_secret flag in case there the secret is empty, it is enough to check op_state.will_gen_secret  (like was done before).
I suggested adding a check in rgw_admin that if the user provides either --secret or --gen-secret in the command line.

guce commented:
user create will gen access by default, gen access will gen secret by default. this two case no need provide --secret or --gen-secret. maybe add the check in access key pool checkop better?

guce commented:
@oritwas 

also need check this
`radosgw-admin key create --uid=111 --access-key=11111`

maybe add the check in RGWAccessKeyPool::generate_key() better ?

guce commented:
@oritwas 

Hi, I have modified, please check.

In `RGWAccessKeyPool::generate_key()`:
* use `gen_secret` which equal to `op_state.will_gen_secret()`
* check if provide `--gen-secret` or `--secret`
* modify judgment about `new_key.subuser` , make subuser can still create s3 key

guce commented:
@oritwas

My latest modified whether is correct?

oritwas commented:
Great!!
